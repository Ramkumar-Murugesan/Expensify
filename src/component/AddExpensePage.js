import React from "react";
import ExpenseForm from "./ExpenseForm";
import { addExpense } from '../actions/expenses';
import { connect } from "react-redux";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.addExpense(expense)
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                this is AddExpensePage component
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>
        )
    }
}

// replace with class 
// const AddExpensePage = (props) => (
//     <div>
//     this is AddExpensePage component
//     <ExpenseForm onSubmit={(expense) => {
//         // replace with connect mapDispatchToProps because below code is difficult to write the test cases
//         // props.dispatch(addExpense(expense));

//         // new one with the help of connect
//         props.addExpense(expense)
//         props.history.push('/');
//     }}/>
//     </div>
// )

const mapStateToProps = undefined;
// name should action generator so changing onSubmit name to addExpense name
const mapDispatchToProps = (dispatch) => {
    return {
        addExpense: (expense) => dispatch(addExpense(expense))
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddExpensePage);