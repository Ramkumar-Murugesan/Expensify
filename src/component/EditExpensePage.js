
import React from "react";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";
import ExpenseForm from "./ExpenseForm";

export class EditExpensePage extends React.Component {
    onEditExpenseHandler = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onRemoveExpenseHandler = () => {
        this.props.removeExpense(this.props.expense.id)
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onEditExpenseHandler}
                />
                <button onClick={this.onRemoveExpenseHandler}>Remove</button>
            </div>
        )
    }
}
// replace with class based component
// const EditExpensePage = (props) => {
//     // get the path variable value from props.match.params or query params from props.location.search and hash value from props.location.hash
//     console.log(props)
//     return (
//         <div>
//             <ExpenseForm
//                 expense={props.expense}
//                 onSubmit={(expense) => {
//                     console.log('update ', expense)
//                     props.dispatch(editExpense(props.expense.id, expense))
//                     props.history.push('/')
//                 }}
//             />
//             <button onClick={() => {
//                 props.dispatch(removeExpense({id: props.expense.id}))
//                 props.history.push('/')
//             }}>Remove</button>
//         </div>
//     )
// }
// access both state and props in the connect (HOC) component
const mapStateToProps = (state, props) => (
    {
        expense: state.expenses.find(expense => expense.id === props.match.params.id)
    }
)
const mapDispatchToProps = (dispatch) => ({
    editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (id) => dispatch(removeExpense({ id: id }))
})
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);