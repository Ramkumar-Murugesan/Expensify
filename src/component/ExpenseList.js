import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectorExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
    <h1>Expense List</h1>
    {
        props.expenses.length === 0 ? (
            <p>No Expenses</p>
        ) : (
            props.expenses.map((expense) => {
                return <ExpenseListItem key={expense.id} {...expense} />
            })
        )
    }
    </div>
)

const mapStateToProps = (state) => {
    return {
        expenses: selectorExpenses(state.expenses, state.filters)
    }
}
const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);
 export default ConnectedExpenseList;











 // step 1

//  const ExpenseList = (props) => (
//     <div>
//     <h1>Expense List</h1>
//     <p>name is {props.name}</p>
//     </div>
// )
// const ConnectedExpenseList = connect((state) => {
//     return {
//         name: 'Ramkumar'
//     }
// })(ExpenseList);
//  export default ConnectedExpenseList;



// step 2

// const ExpenseList = (props) => (
//     <div>
//     <h1>Expense List</h1>
//     <p>filter Text is {props.filters.text}</p>
//     <p>length is {props.expenses.length}</p>
//     </div>
// )
// const ConnectedExpenseList = connect((state) => {
//     return {
//         expenses: state.expenses,
//         filters: state.filters
//     }
// })(ExpenseList);
//  export default ConnectedExpenseList;


// step 3

// const ExpenseList = (props) => (
//     <div>
//     <h1>Expense List</h1>
//     {
//         props.expenses.map((expense) => {
//             return <ExpenseListItem key={expense.id} {...expense} />
//         })
//     }
//     </div>
// )

// const mapStateToProps = (state) => {
//     return {
//         expenses: state.expenses,
//         filters: state.filters
//     }
// }
// const ConnectedExpenseList = connect(mapStateToProps)(ExpenseList);
//  export default ConnectedExpenseList;

