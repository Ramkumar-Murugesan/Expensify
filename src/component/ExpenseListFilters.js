import React from "react";
import { connect } from "react-redux";
import { setTextFilter, sortByAmount, sortByDate, setStartDate, setEndDate } from "../actions/filters";
import { DateRangePicker } from 'react-dates';

export class ExpenseListFilters extends React.Component {
    state = {
        calenderFocused: null
    }
    onDatesChange = ({ startDate, endDate }) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    }
    onFocusChange = (calenderFocused) => {
        this.setState(() => ({ calenderFocused }));
    }
    onTextFilter = (e) => {
        this.props.setTextFilter(e.target.value);
        console.log(e.target.value)
    }
    onSortChange = (e) => {
        e.target.value == 'amount' ? this.props.sortByAmount() : this.props.sortByDate()
    }
    render() {
        return (
            <div>
                <input type="text" value={this.props.filters.text} onChange={this.onTextFilter} />
                <select value={this.props.filters.sortBy} onChange={this.onSortChange}>
                    <option value="date">date</option>
                    <option value="amount">amount</option>
                </select>
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    endDate={this.props.filters.endDate}
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calenderFocused}
                    onFocusChange={this.onFocusChange}
                    showClearDates={true}
                    numberOfMonths={1}
                    isOutsideRange={() => false}
                />
            </div>
        )
    }
}


const mapStateToProps = (state) => {
    return {
        filters: state.filters
    }
}
const mapDispatchToProps = (dispatch) => ({
    setStartDate: (startDate) => dispatch(setStartDate(startDate)),
    setEndDate  : (endDate) => dispatch(setEndDate(endDate)),
    setTextFilter: (text) => dispatch(setTextFilter(text)),
    sortByAmount: () => dispatch(sortByAmount()),
    sortByDate: () => dispatch(sortByDate())
})

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);