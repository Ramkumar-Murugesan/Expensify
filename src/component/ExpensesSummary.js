import React from "react";
import { connect } from "react-redux";
import expensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import numeral from "numeral";

export const ExpensesSummary = (props) => {
    const expenseWord = props.expensesCount == 1 ? 'expense' : 'expenses'
    return (
        <div>
        Viewing {props.expensesCount} {expenseWord} totalling  {numeral(props.expensesTotal / 100).format('$0,0.00')}
    </div>
    )
}

const mapStateToProps = (state) => {
    const visibleExpenses = selectExpenses(state.expenses, state.filters)
   return {
        expensesCount: visibleExpenses.length,
        expensesTotal: expensesTotal(visibleExpenses)
    }
}

export default connect(mapStateToProps)(ExpensesSummary)