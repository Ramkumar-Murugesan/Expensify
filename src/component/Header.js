
import React from "react";
import { Link } from 'react-router-dom';
import { connect } from "react-redux";
import { startLogout } from "../actions/auth";

export const Header = ({ startLogout }) => (
  <div className="header">
    <div className="content-container">
      <div className="header__content">
        <Link className="header__title" to="/dashboard">
          <h1>Expensify</h1>
        </Link>
        <button className="button button--link" onClick={startLogout}>Logout</button>
      </div>
    </div>
  </div>
)




const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout())
});

export default connect(undefined, mapDispatchToProps)(Header);



   // error -- missed () in startLogout function
// const mapDispatchToProps = (dispatch) => ({
//     startLogout: () => dispatch(startLogout)
// })

//    export default connect(undefined, mapDispatchToProps)(Header);






// before applying style without removing anything
// export const Header = ({startLogout}) => (
//   <div>
//   <h1>Expensify</h1>
//   <div>
//   <NavLink to="/dashboard" activeClassName='is-active'>Dashboard</NavLink>
//   <NavLink to="/create" activeClassName='is-active'>Create Expense</NavLink>
//   <NavLink to="/edit/1" activeClassName='is-active'>Edit Expense</NavLink>
//   <NavLink to="/help" activeClassName='is-active'>Help</NavLink>
//   <button onClick={startLogout}>Logout</button>
//   </div>
//   </div>
//  )