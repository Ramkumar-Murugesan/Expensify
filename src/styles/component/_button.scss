.button {
    padding: $s-size;
    background-color: $blue;
    border: none;
    font-weight: 300;
    font-size: $font-size-large;
    display: inline-block;
    text-decoration: none;
    line-height: 1;
    color: white;
    cursor: pointer;
}

.button--link {
    background: none;
}

.button--secondary {
    background: #888888;
}