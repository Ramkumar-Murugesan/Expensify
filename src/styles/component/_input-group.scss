.input-group {
    // style for mobile screens
    display: flex;
    flex-direction: column;
    margin-bottom: $m-size;
    
    // style for larger screens ie desktop
    @media (min-width: $desktop-breakpoint) {
        flex-direction: row;
        margin-bottom: $l-size;
    }
}

.input-group__item {
 margin-bottom: $s-size;

 // style for larger screens ie desktop
 @media (min-width: $desktop-breakpoint) {
    margin: 0 $s-size 0 0;
}
}