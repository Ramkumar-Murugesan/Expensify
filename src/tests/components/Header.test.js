import React from "react";

// no longer needed because we are using enzyme shallow render
// import ReactShallowRenderer from 'react-test-renderer/shallow';

import { shallow } from 'enzyme';
// add globally for our application in jest.config.json --> so no need to import every file. it automatically convert it
// import toJSON from 'enzyme-to-json';
import {Header} from '../../component/Header';


// enzyme shallow renderer example
test('should render header correctly', () => {
    const wrapper = shallow(<Header />);
    expect(wrapper).toMatchSnapshot()


    // expect(toJSON(wrapper)).toMatchSnapshot()
    // expect(wrapper.find('h1').text()).toBe("Expensify")
})


test('should call startLogout on button click', () => {
    const startLogout = jest.fn();
    const wrapper = shallow(<Header startLogout={startLogout} />);
    wrapper.find('button').simulate('click');
    expect(startLogout).toHaveBeenCalled();
  });

// React shallow renderer example
// test('should render header correctly', () => {
//     const renderer = new ReactShallowRenderer();
//     // add react component to render
//     renderer.render(<Header />)
//     // should show the rendered component with props 
//     // console.log('shallow renderee ', renderer.getRenderOutput())

//     // we does not pass any value toMatchSnapshot method then it automatically take the snapshot(make copy of component) and store it in our application
//     // 2nd time when we test the application the jest use the taken snapshot and check with our render component
//     // if it match then the test pass  === if it not then the test is failed
//     // to update the snapshot of what we change in render component. then press u in terminal
//     expect(renderer.getRenderOutput()).toMatchSnapshot();
// })