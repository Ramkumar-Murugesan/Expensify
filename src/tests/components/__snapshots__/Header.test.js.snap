// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render header correctly 1`] = `
<div>
  <h1>
    Expensify
  </h1>
  <div>
    <NavLink
      activeClassName="is-active"
      aria-current="page"
      exact={true}
      to="/"
    >
      Dashboard
    </NavLink>
    <NavLink
      activeClassName="is-active"
      aria-current="page"
      to="/create"
    >
      Create Expense
    </NavLink>
    <NavLink
      activeClassName="is-active"
      aria-current="page"
      to="/edit/1"
    >
      Edit Expense
    </NavLink>
    <NavLink
      activeClassName="is-active"
      aria-current="page"
      to="/help"
    >
      Help
    </NavLink>
  </div>
</div>
`;
